File: ios/StudyApp/Data/Sync/SyncRevisionStamper.swift
import Foundation
import CryptoKit

/// Stamps revision metadata onto outbound envelopes. Optional fields remain
/// absent when unchanged so legacy clients can still read documents.
enum SyncRevisionStamper {
    static func stamp(
        _ envelopes: [SyncEntityEnvelope],
        previousBase: AppData?,
        previousRevisions: [String: String] = [:],
        deviceId: String = SyncDeviceIdentity.current
    ) -> [SyncEntityEnvelope] {
        let previousRevisions = previousRevisions.isEmpty ? previousRevisionMap(from: previousBase) : previousRevisions
        return envelopes.map { envelope in
            var stamped = envelope
            let parent = previousRevisions[envelope.documentId]
            stamped.revisionId = UUID().uuidString.lowercased()
            stamped.parentRevisionId = parent
            stamped.deviceId = deviceId
            stamped.contentHash = sha256(envelope.json)
            return stamped
        }
    }

    private static func previousRevisionMap(from base: AppData?) -> [String: String] {
        guard let base else { return [:] }
        return Dictionary(
            uniqueKeysWithValues: SyncDeltaSerializer.decompose(base).compactMap { envelope in
                guard let revision = envelope.revisionId ?? envelope.contentHash, !revision.isEmpty else {
                    return nil
                }
                return (envelope.documentId, revision)
            }
        )
    }

    private static func sha256(_ json: String) -> String {
        let digest = SHA256.hash(data: Data(json.utf8))
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}

File: android/app/src/main/java/com/studyapp/sync/SyncRevisionStamper.kt
package com.studyapp.sync

import com.studyapp.domain.usecase.AppData
import java.security.MessageDigest
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SyncRevisionStamper @Inject constructor(
    private val deviceIdentity: SyncDeviceIdentity
) {
    fun stamp(
        envelopes: List<SyncEntityEnvelope>,
        previousBase: AppData?,
        previousRevisions: Map<String, String> = previousRevisionMap(previousBase)
    ): List<SyncEntityEnvelope> {
        val deviceId = deviceIdentity.current
        return envelopes.map { envelope ->
            envelope.copy(
                revisionId = UUID.randomUUID().toString().lowercase(),
                parentRevisionId = previousRevisions[envelope.documentId],
                deviceId = deviceId,
                contentHash = sha256(envelope.json)
            )
        }
    }

    private fun previousRevisionMap(base: AppData?): Map<String, String> {
        if (base == null) return emptyMap()
        return SyncDeltaSerializer.decompose(base).mapNotNull { envelope ->
            val revision = envelope.revisionId ?: envelope.contentHash
            if (revision.isNullOrEmpty()) null else envelope.documentId to revision
        }.toMap()
    }

    private fun sha256(json: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(json.toByteArray())
        return digest.joinToString("") { "%02x".format(it) }
    }
}
